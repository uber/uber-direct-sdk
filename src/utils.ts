import { Error } from "./deliveries/types";

type Headers = {
  Authorization: string;
  'Content-Type'?: string;
};


type ApiError = {
  code?: string;
  message: string;
  kind?: string;
  metadata?: Record<string, unknown>;
  status?: number;
}

export class FetchError extends Error {
  public status: number;
  public code?: string;
  public metadata?: Record<string, unknown>;

  constructor(message: string, status: number, code?: string, metadata?: Record<string, unknown>) {
    super(message);
    this.status = status;
    this.code = code;
    this.metadata = metadata;
  }
}

const getHeaders = (accessToken: string, method: 'GET' | 'POST'): Headers => {
  const headers: Headers = {
    Authorization: `Bearer ${accessToken}`,
  };

  if (method === 'POST') {
    headers['Content-Type'] = 'application/json';
  }

  return headers;
};

const makeQueryString = (params: Record<string, unknown>) => {
  const esc = encodeURIComponent;
  const query = Object.keys(params)
    .map((k) => `${esc(k)}=${esc(params[k] as string)}`)
    .join('&');
  return `?${query}`;
};


const fetchData = async <T = any>(
  url: string,
  method: 'GET' | 'POST',
  accessToken: string,
  req?: Record<string, unknown>
) => {
  const headers: Record<string, string> = {
    ...getHeaders(accessToken, method),
    'Content-Type': 'application/json',
  };

  const options: RequestInit = {
    method: method,
    headers: headers,
    body: req ? JSON.stringify(req) : undefined,
  };

  try {
    const response = await fetch(url, options);
    if (response.ok) {
      const data = await response.json();
      return data;
    } else {
      const errorResponse: ApiError = await response.json();
      throw new FetchError(
        errorResponse.message,
        response.status,
        errorResponse.code,
        errorResponse.metadata
      );
    }
  } catch (error) {
    if (error instanceof FetchError) {
      // If it's our custom error, rethrow it
      throw error;
    } else {
      // For other errors, wrap them in our custom error class
      throw new FetchError(error as string, 500);
    }
  }
};

export { fetchData, getHeaders, makeQueryString };
