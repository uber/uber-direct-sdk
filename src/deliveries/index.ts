import type {
  DeliveryQuoteReq,
  DeliveryReq,
  ListDeliveriesReq,
  PODReq,
  UpdateDeliveryReq,
} from './types';
import { fetchData, makeQueryString } from '../utils';

export class DeliveriesClient {
  accessToken: string;
  baseURL: string;

  constructor(accessToken: string) {
    this.accessToken = accessToken;

    const customerId = process.env.UBER_DIRECT_CUSTOMER_ID;
    if (!customerId) {
      throw new Error('Must include UBER_DIRECT_CUSTOMER_ID in environment variables');
    }
    this.baseURL = `https://api.uber.com/v1/customers/${customerId}`;
  }

  async createQuote(req: DeliveryQuoteReq) {
    return fetchData(`${this.baseURL}/delivery_quotes`, 'POST', this.accessToken, req);
  }

  async createDelivery(req: DeliveryReq) {
    return fetchData(`${this.baseURL}/deliveries`, 'POST', this.accessToken, req);
  }

  async getDelivery(deliveryId: string) {
    return fetchData(`${this.baseURL}/deliveries/${deliveryId}`, 'GET', this.accessToken);
  }

  async listDeliveries(options?: ListDeliveriesReq) {
    let url = `${this.baseURL}/deliveries`;
    if (options) {
      url += makeQueryString(options);
    }
    return fetchData(url, 'GET', this.accessToken);
  }

  async cancelDelivery(deliveryId: string) {
    return fetchData(`${this.baseURL}/deliveries/${deliveryId}/cancel`, 'POST', this.accessToken);
  }

  async updateDelivery(deliveryId: string, req: UpdateDeliveryReq) {
    return fetchData(`${this.baseURL}/deliveries/${deliveryId}`, 'POST', this.accessToken, req);
  }

  async proofOfDelivery(deliveryId: string, req: PODReq) {
    return fetchData(
      `${this.baseURL}/deliveries/${deliveryId}/proof-of-delivery`,
      'POST',
      this.accessToken,
      req
    );
  }
}

export const createDeliveriesClient = (accessToken: string) => {
  return new DeliveriesClient(accessToken);
};
