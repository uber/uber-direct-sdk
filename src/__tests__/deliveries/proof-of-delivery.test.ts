import { DeliveriesClient, createDeliveriesClient } from '../../deliveries';
import {
  podReq,
  podResp,
  accessToken,
  customerId,
  deliveryId,
  invalidPODReq,
} from '../../__fixtures__/index';
import { getHeaders } from '../../utils';

describe('proofOfDelivery', () => {
  let deliveriesClient: DeliveriesClient;
  beforeEach(() => {
    jest.resetModules();

    process.env.UBER_DIRECT_CUSTOMER_ID = customerId;
    deliveriesClient = createDeliveriesClient(accessToken);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should make a POST request with the required parameters', async () => {
    global.fetch = jest.fn().mockResolvedValueOnce({
      ok: true,
      json: jest.fn().mockResolvedValueOnce(podResp),
    });

    const pod = await deliveriesClient.proofOfDelivery(deliveryId, podReq);
    const method = 'POST';
    expect(global.fetch).toHaveBeenCalledWith(
      `${deliveriesClient.baseURL}/deliveries/${deliveryId}/proof-of-delivery`,
      {
        method,
        headers: getHeaders(accessToken, method),
        body: JSON.stringify(podReq),
      }
    );
    expect(pod).toEqual(podResp);
  });

  it('should throw an error if response.ok is not true', async () => {
    global.fetch = jest.fn().mockResolvedValueOnce({
      ok: false,
    });

    await expect(
      deliveriesClient.proofOfDelivery(deliveryId, podReq)
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it('should handle errors', async () => {
    global.fetch = jest.fn().mockRejectedValueOnce({
      response: {
        data: {
          code: 'invalid_params',
          message: 'The parameters of your request were invalid.',
          kind: 'error',
          metadata: {
            pickup_address: 'This field is required.',
          },
        },
      },
    });

    await expect(
      deliveriesClient.proofOfDelivery(deliveryId, invalidPODReq)
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});
