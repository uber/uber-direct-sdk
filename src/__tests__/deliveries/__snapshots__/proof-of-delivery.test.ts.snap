// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`proofOfDelivery should handle errors 1`] = `"[object Object]"`;

exports[`proofOfDelivery should throw an error if response.ok is not true 1`] = `"TypeError: response.json is not a function"`;
